### ISSUE TITLE
pg-cdc: Schema changes not handled, wrong data returned

### ISSUE THREAD

**Issue by philip-stoev** [MaterializeInc/materialize#6739]
### What version of Materialize are you using?

```
materialized v0.7.4-dev (adc7083b1)
```

### What was the issue?

DDL is not handled at all by the Postgres CDC. While this is a known limitation, I am opening a ticket so that it is tracked properly.

The original design notes in #5370 say "it's unclear how to handle upstream schema changes. I believe this is an existing limitation with debezium as well." however strictly speaking this is not the case. Debezium will either transparently handle certain compatible schema changes or it will report an error its log and there is an opportunity for Mz to perform some action, such as erroring the source.
 
### Is the issue reproducible? If so, please provide reproduction instructions.

I checked the following cases:

1. ```DROP TABLE```

Nothing happens on the Mz side. Nothing in the logs.

2. ```ALTER TABLE```

This one is the more problematic scenario. If one does the following on the Postgres side:

```

drop table if exists t1;
create table t1 (f1 integer, f2 integer, f3 integer);
insert into t1 values (1,2,3);
alter table t1 add column f4 integer;
insert into t1 values (1,2,3,4);
alter table t1 drop column f1;
insert into t1 values (2,3,4);
select * from t1;

postgres=# select * from t1;
 f2 | f3 | f4 
----+----+----
  2 |  3 |   
  2 |  3 |  4
  2 |  3 |  4
```

The Mz side will look like this:

```
materialize=> CREATE VIEWS FROM SOURCE "mz_source" (t1);
CREATE VIEW
materialize=> select * from t1;
 f2 | f3 | f4 
----+----+----
  1 |  2 |  3
  1 |  2 |  3
  2 |  3 |  4
(3 rows)
```

**Comment by philip-stoev** [comment]
The view is created with individual columns referenced positionally , e.g.

```
materialize=> show create view t1;
         View          |                                                                            Create View                                                                            
-----------------------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------
 materialize.public.t1 | CREATE VIEW "materialize"."public"."t1" ("f1") AS SELECT ("row_data"[1])::"pg_catalog"."int4" AS "f1" FROM "materialize"."public"."mz_source" WHERE "oid" = 16386
(1 row)
```

But column 1 in the array may mean one thing in one row of the source and another in another row.

Furthemore, if the index is out of bounds, NULL will silently be returned:

```
materialize=> SELECT ("row_data"[2222])::"pg_catalog"."int4" AS "f1" FROM "materialize"."public"."mz_source" WHERE "oid" = 16386;
 f1 
----
   
(1 row)
```

**Comment by awang** [comment]
Saving what we quickly discussed for greater visibility + discussion.  Some of our options for how to proceed are:

(1) ignore the alter
This leads to wrong data and makes the direct pg replication more fragile compared to debezium in that respect as debezium will stop replicating on unsafe DDLs in most cases.

(2) drop+completely re-create the table
We hope that this is doable in the form of erroring the source until it can be recreated. All the downstream views will also need to be dropped, recreated and everything will be rehydrated with data from scratch

**Comment by philip-stoev** [comment]
The Postgresql documentation mentions limitations on logical replication + DDL:

https://www.postgresql.org/docs/13/logical-replication-restrictions.html

At the same time, Debezium is able to detect TRUNCATE and other DDL events and act upon them -- TRUNCATE events are sent to the kafka topic and DDLs cause the new schema to be registered in the Schema Registry. I do not know how this is achieved.

**Comment by elindsey** [comment]
Logical decoding plugins do not export ddl statements, but as of v10.0 iirc there is support for communicating the schema on change and I believe the default `pgoutput` takes advantage of this. Retitled the task to reflect this nuance.

**Comment by nharring-adjacent** [comment]
Since this has turned into a bit of a rabbit hole I am going to add everything I have figured out and/or tried so far to make it easier to share context.

- We can detect ALTER TABLE queries by watching for `RELATION` messages and then comparing the described schema to what we have cached for that relation_id. (https://www.postgresql.org/docs/10/protocol-logical-replication.html#PROTOCOL-LOGICAL-MESSAGES-FLOW indicates that they should only come the first time a relation_id has transactions being sent for it or if it changes but there is not much clarity around how it determines first and in a little experimenting I could not accurately determine if something had changed unless I diffed the actual schema)
- Because we handle replication for all tables from a given source with a single stream we want to minimize the number of alters (or truncates, see [6812](https://github.com/MaterializeInc/materialize/issues/6812)) that would cause the source to error out it would be ideal to capture the set of tables that were known to exist when the source was materialized and ignore events which trigger errors for any other table since it would be expected they are not to be part of any dataflows.
  - We could do this when we begin reading but this means if `materialized` is restarted after tables are added to an upstream publication then they would expand the scope of error sources after the restart. Petros pointed me to [this todo](https://github.com/MaterializeInc/materialize/blob/main/src/sql/src/pure.rs#L199-L200) where he planned to capture this upstream schema information during purification of the `CREATE MATERIALIZED SOURCE`.
  - During purification the current implementation queries the publication and returns a `Vec<postgres_util::TableInfo>` however `postgres_util` crate depends on `sql_parser` so the `sql_parser::ast::CreateSourceConnector::Postgres` type cannot use the `Vec<TableInfo>`. My initial plan was to convert `postgres_util::TableInfo` into `sql_parser::ast::PgTable<T: AstInfo>` as `sql_parser::ast::PgTable<Raw>` however because these tables are defined in the remote catalog this leads to errors like `ERROR:  unknown catalog item 'pk_table'` when trying to create a materialized source.
- Regardless of format in the catalog the initialization logic for the `PostgresSourceReader` will need to using the metadata from the catalog which will be compared against what the publication reports when the reader is actually instantiated and schema changes or tables disappearing can be turned into errors. 
  - This poses a backwards compatibility dilemma with the following options: 
    - gracefully handle the catalog not having the metadata (forever? can we update the catalog at this stage the first time it happens?)
    - document a breaking change and force all materialized sources to be recreated
    - write a migration which uses the current catalog information to read the remote publication and add the metadata to the catalog
- During init the metadata will also be stored on the `PostgresSourceReader` struct for runtime use:
  - When `RELATION` messages arrive they will be checked against the "relevant" list of relation ids, if the id is relevant then the schema will be diffed against what was in the catalog and if its different the source will go into an error state referencing the table which had an ALTER run against it
  - When `TRUNCATE` messages arrive we will compute the intersection of the truncated tables and relevant tables, if this set has any members the source will go into an error state referencing the set of tables which had TRUNCATE run

**Comment by petrosagg** [comment]
> This poses a backwards compatibility dilemma with the following options: 

We have done a similar migration to the past with CSV sources that in the beginning did not store the schema observed on creation in the catalog but we later wanted to capture that information. This is the migration: https://github.com/MaterializeInc/materialize/blob/main/src/coord/src/catalog/migrate.rs#L617-L632

Granted, talking to an upstream postgres database is not the same as re-reading a local file but I think it is not unreasonable to do that during a version upgrade, especially if it saves us the headache of always thinking "oh, is this pg source with or without metadata?" when a customer comes with an issue. @benesch does this match your expectation?

> however because these tables are defined in the remote catalog this leads to errors

It looks like `sql_parser::ast::PgTable` is dead code at the moment, must be left over from some previous iteration. Also, it doesn't quite capture what we need. For example it has an `alias` field that the postgres source doesn't really care about.

In order to not have name resolution errors you need to make sure that you use `T::ObjectInfo` (`where T: AstInfo`) only for objects that exists in our catalog. If you want to store something that is still an object name but it's from the remote pg catalog then you should use `UnresolvedObjectName` directly.

I think the definition of `PgTable` that we need is:

```rust
struct PgTable<T: AstInfo> {
    oid: u64, // or whatever the integer type of oid is, can't remember now
    name: UnresovledObjectName,
    columns: Vec<ColumnDef<T>,
}
```

Btw, when we do the above we can also change `CREATE VIEWS FROM SOURCE` to use the information in our catalog instead of reaching out to the upstream database. I'll create an issue about this

**Comment by nharring-adjacent** [comment]
It might be good to roll that change into the work for this issue, otherwise we risk someone creating the materialized source, a table being added to the upstream catalog, then doing `CREATE VIEWS FROM SOURCE` which would result in a view being created against a table which isn't in the materialize catalog metadata.

**Comment by nharring-adjacent** [comment]
Two questions about putting the oid into the `PgTable` struct:
1. The compile of sql-parser breaks when I try to use `u32` as the field type with an error like:
```
error: failed to run custom build command for `sql-parser v0.0.0 (/home/nharring/materialize/src/sql-parser)`

Caused by:
  process didn't exit successfully: `/home/nharring/materialize/target/debug/build/sql-parser-92ec556f31349363/build-script-build` (exit status: 1)
  --- stderr
  Error: Unable to analyze non built-in type that is not defined in input: u32
```
It works if I store it as a `String` which seems like not a big deal but it'd be great not to have to convert it back and forth as much.
2. How would we want to represent this in the AstFormatter? Currently we write:
```
tbl1_name (col1_name col1_type, col2_name col2_type,...), tbl2_name (col1_name col1_type, ...)
```
The most obvious way to me to add OID without adding more keywords would be
```
tbl1_oid AS tbl1_name (col1_name col1_type, col2_name col2_type,...), tbl2_oid as tbl2_name (col1_name col1_type,...)
```

**Comment by petrosagg** [comment]
Regarding your first question, this is where we define the list of allowed AST types https://github.com/MaterializeInc/materialize/blob/main/src/walkabout/src/ir.rs#L304-L310 . If you add `u32` there it should work fine.

Regarding 2. another option is `tbl1_name (col1_name col1_type, col2_name col2_type,...) WITH (oid = 321)`

**Comment by nharring-adjacent** [comment]
This is resolved as of #10721 being merged
